### ISSUE TITLE
Inconsistent mapping in multiple Talon runs

### ISSUE THREAD

**Issue by kwells4** [mortazavilab/TALON#103]
# Description of problem

Thanks for writing this great tool!

I've been trying to run `Talon` on a promethion run with 3 samples. Overall, I have about 100G of data but am mostly running into trouble because there is about 42G of data on chromosome 7 (Insulin in the pancreas is a bit of a problem).

To get around this, I am running `Talon` in pieces as suggested in #87 and using the development branch to use bam as input as suggested in #76.

I'm first mapping with `minimap2` and then running the `Talon` pipeline.

To make sure this is working, I first ran `Talon` on a small subset of my data and then tested running `Talon` in pieces on the same small subset of my data.

The problem is the two do not give the exact same answer. For example, running `Talon` on the full dataset combined, the talon_read_annot.tsv had this information for one read
```
e6910aaf-809f-45cf-9ffb-3dddba64950f    WT_mouse3   GRCm38  15  35931980    35938209    +   4   408 45631   167125  ENSMUSG00000037646  GRCm38_novelT000167125  Vps13b  GRCm38_novelT000167125  Known   NNC None    0.25    None    None    None    None
```

While running it on a subset gave this output:
```
e6910aaf-809f-45cf-9ffb-3dddba64950f    WT_mouse3_0013  GRCm38  15  35931980    35938159    +   4   408 55606   142012  GRCm38_novelG000055606  GRCm38_novelT000142012  GRCm38_novelG000055606  GRCm38_novelT000142012  Antisense   NNC None    0.25    None    None    None    None
```

Here, one is labeled as mapping to a gene while the other is labeled as "Antisense". Additionally, the start and end positions are not the same. When looking at the bam file using pysam, the start position is 35931979 and the end position is 35938159 which more closely agrees with the second output. The reads are identical in the bam file that is provided to `Talon`. Do you know why there is a discrepancy between the two methods and why the annotation file doesn't have the correct read length?

# Overview of steps

I first split up my fastq files into 18 smaller fastq files to run minimap and internal priming in a parallel fashion. This is all done in a `Snakemake` pipeline, but I will translate here.

## Subset fastq files into smaller pieces
```bash
fastp -i {sample_fastq} \
    -Q -A --thread 1 --split_prefix_digits=4 \
    --out1={sample_name}_split.fastq.gz \
    --split=18
```

## Minimap on all split files

```bash
minimap2 -ax splice \
    -uf --MD --sam-hit-only \
    --junc-bed genome_junctions.bed \
    genome_ref.mmi \
    {id}.{sample_name}_split.fastq.gz > {sample}.{id}.aligned.sam

samtools view -Sb {sample}.{id}.aligned.sam -o {sample}.{id}.unsorted.bam
samtools sort {sample}.{id}.unsorted.bam -o {sample}.{id}.sorted.bam
samtools index {sample}.{id}.sorted.bam
```

## Database creation
```bash
prefix=GRCm38_novel
annotation=GRCm38_annot

talon_initialize_database \
    --f Mus_musculus.GRCm38.96.gtf \
    --a $annotation \
    --g GRCm38 \
    --idprefix $prefix \
    --o talon_database
```

## Internal priming
```bash 
samfile={sample}.{id}.internal_priming_labeled.sam

talon_label_reads \
        --f {sample}.{id}.sorted.bam \
        --g Mus_musculus.GRCm38.dna.primary_assembly.fa \
        --ar 20 \
        --deleteTmp \
        --o {sample}.{id}.internal_priming

samtools view -Sb $samfile -o {output}
```

## Talon
When running `Talon` on everything combined, I first combined all of my 18 bam files for each sample and added these combined bam files to the config file (one per line). When running `Talon` on each separately, I made one config file per bam file and ran `Talon` individually for each bam file.

```bash
talon \
    --f config.tsv \
    --db talon_database.db \
    --build GRCm38 \
    --cov 0.65 \
    --identity 0.8 \
    --threads 20 \
    --o talon_{id}_out
```

**Comment by fairliereese** [comment]
Hi there,

I'm pretty puzzled by what you're seeing as well, but I think there might be explanations for some of the inconsistencies. I think your best bet to diagnose what's going on is to make a genome browser plot of your read and the transcript models that are called by each of your TALON runs (ie for the subsetting approach and the full dataset approach).

TALON does exhibit some stochasticity based on what order each read is processed in which can lead to differential assignments to genes or transcripts. So basically I would be interested in seeing what transcripts are being called in that region where the read is to try to explain the behavior. I would want to see this before trying to jump to conclusions about what exactly is happening. 

I am, however, deeply confused about why the read ends are not being reported as the same genomic coordinates if the entries in the bam file are identical. I'm not sure I have an explanation for that at the moment, but perhaps visualizing the data will also yield some insight.

**Comment by kwells4** [comment]
Thanks for the detailed and fast respose!

I was wondering if the order reads were processed would change the results. I did notice some of that behavior where the transcripts were annotated differently in my different runs.

For this specific example, I have plotted three tracks, one from each of the bam files used in my processing pipeline. All tracks look identical and actually map to the Cox6c gene better than the Vps13b gene. These do all map on the + strand while Cox6c is on the - strand.
![inconsistent_read](https://user-images.githubusercontent.com/14188899/165652393-46e3db12-68d9-43a0-8ae8-a3b4a4eaeec1.png)

Interestingly, all the same reads were grouped together in both runs and all of the reads have a very similar alignment profile.
![combined_combined](https://user-images.githubusercontent.com/14188899/165652728-df553b73-50a6-4b9a-a942-1642651129d1.png)

Running `infer_experiment.py` on my data, it does seem like my experiment is unstranded. I had put that on the back-burner to deal with once the rest of my pipeline was developed, but these problems may be related so I will start working on that now. I didn't see a way to make `Talon` work with unstranded data, so I'm thinking of using `bedtools intersect` to essentially get around the problem, unless you have a better solution. I'll close this for now while I try to see if the strandedness is somehow causing the bug.
